pkg/testutils/integration_test.go

```diff
@@ -0,0 +1,106 @@
+package testutils_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/Azure/adx-mon/pkg/testutils"
+	"github.com/Azure/adx-mon/pkg/testutils/collector"
+	"github.com/Azure/adx-mon/pkg/testutils/ingestor"
+	"github.com/Azure/adx-mon/pkg/testutils/kustainer"
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/modules/k3s"
+)
+
+func TestIntegration(t *testing.T) {
+	// An extra generous timeout for the test. The test should run in
+	// about 5 minutes, but when running with the race detector, it
+	// can take longer.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Hour)
+	defer cancel()
+
+	k3sContainer, err := k3s.Run(ctx, "rancher/k3s:v1.31.2-k3s1")
+	testcontainers.CleanupContainer(t, k3sContainer)
+	require.NoError(t, err)
+
+	kustoContainer, err := kustainer.Run(ctx, "mcr.microsoft.com/azuredataexplorer/kustainer-linux:latest", kustainer.WithCluster(ctx, k3sContainer))
+	testcontainers.CleanupContainer(t, kustoContainer)
+	require.NoError(t, err)
+
+	restConfig, _, err := testutils.GetKubeConfig(ctx, k3sContainer)
+	require.NoError(t, err)
+	require.NoError(t, kustoContainer.PortForward(ctx, restConfig))
+
+	kubeconfig, err := testutils.WriteKubeConfig(ctx, k3sContainer, t.TempDir())
+	require.NoError(t, err)
+	t.Logf("Kubeconfig: %s", kubeconfig)
+	t.Logf("Kustainer: %s", kustoContainer.ConnectionUrl())
+
+	t.Run("Create databases", func(t *testing.T) {
+		for _, dbName := range []string{"Metrics", "Logs"} {
+			require.NoError(t, kustoContainer.CreateDatabase(ctx, dbName))
+		}
+	})
+
+	ingestorContainer, err := ingestor.Run(ctx, ingestor.WithCluster(ctx, k3sContainer))
+	testcontainers.CleanupContainer(t, ingestorContainer)
+	require.NoError(t, err)
+
+	collectorContainer, err := collector.Run(ctx, collector.WithCluster(ctx, k3sContainer))
+	testcontainers.CleanupContainer(t, collectorContainer)
+	require.NoError(t, err)
+
+	t.Run("Logs", func(t *testing.T) {
```

Comment:
Can we set the Logs and Metrics blocks to run with `t.Parallel()` so they can wait in parallel?

Reply:
Everything that can be ran in parallel is now doing so. I decided to use go routines to execute blocks in parallel instead of `t.Parallel()` due to difficulties with the way Go executes parallel tests.